apps/prairielearn/src/pages/shared/syncHelpers.js

```diff
@@ -53,48 +51,43 @@ export async function gitStatus(locals) {
   return serverJob.jobSequenceId;
 }
 
-function locateImage(image, callback) {
-  docker.listImages(function (err, list = []) {
-    if (ERR(err, callback)) return;
-    for (let i = 0, len = list.length; i < len; i++) {
-      if (list[i].RepoTags && list[i].RepoTags?.indexOf(image) !== -1) {
-        return callback(null, docker.getImage(list[i].Id));
-      }
-    }
-    return callback(new Error(`Unable to find image=${image}`));
-  });
-}
-
-function confirmOrCreateECRRepo(repo, job, callback) {
+/**
+ * @param {string} repo
+ * @param {import('../../lib/server-jobs').ServerJob} job
+ */
+async function ensureECRRepo(repo, job) {
   const ecr = new ECR(makeAwsClientConfig());
   job.info(`Describing repositories with name: ${repo}`);
-  ecr.describeRepositories({ repositoryNames: [repo] }, (err, data) => {
-    let repositoryFound = false;
-    if (err) {
-      job.info(`Error returned from describeRepositories(): ${err}`);
-      job.info('Treating this error as meaning the desired repository does not exist');
-    } else {
-      repositoryFound = !!_.find(data.repositories, ['repositoryName', repo]);
-    }
 
-    if (!repositoryFound) {
-      job.info('Repository not found');
-
-      job.info(`Creating repository: ${repo}`);
-      const params = {
-        repositoryName: repo,
-      };
-      ecr.createRepository(params, (err) => {
-        if (ERR(err, callback)) return;
-        job.info('Successfully created repository');
-        callback(null);
-      });
+  let repositoryFound = false;
+  try {
+    const data = await ecr.describeRepositories({ repositoryNames: [repo] });
+    repositoryFound = !!data.repositories?.find((r) => r.repositoryName === repo);
+  } catch (err) {
+    if (err instanceof RepositoryNotFoundException) {
+      // Repository not found; this is expected.
     } else {
-      job.info('Repository found');
-      // Already exists, nothing to do
-      callback(null);
+      // Something else went wrong; allow it to bubble up.
+      Sentry.captureException(err, {
+        tags: {
+          repository: repo,
+        },
+      });
+      throw err;
     }
-  });
+  }
+
+  if (!repositoryFound) {
+    job.info('Repository not found');
+
+    job.info(`Creating repository: ${repo}`);
+    await ecr.createRepository({ repositoryName: repo });
```

Comment:
Not that it really matters, but we could also catch `RepositoryNotFoundException` here and continue. This would avoid the race condition where two processes are creating the repository simultaneously.

Reply:
Hmm... I sorta like where that's going, maybe instead we could just _always_ try to create the repository instead of first checking if it exists? Then we have unified handling for both race conditions and the general situation where the repository already exists.

Reply:
That would certainly be clean code. I have a slight concern that we'll hit some edge case in AWS APIs where a failed creation is a lot slower than a check for existence, or something like that, but I obviously don't have any data to support this worry. Go for it if you want!

Reply:
In 0b89b9d, I went with catching `RepositoryAlreadyExistsException` (note that this is different from `RepositoryNotFoundException`). Take another look and let me know if this looks good!